### ISSUE TITLE
CupertinoTextField should call onChanged when back button is pressed on empty text

### ISSUE THREAD

**Issue by fralways** [flutter/flutter#42598]
Appropriate `TextField` delegate on iOS gets called after every user input, not only when new value is different than old value which it seems how it works in Flutter. This should allow fine user input handling which is needed in special cases. For example, if we have multiple text fields in a row, and user populates one and app auto focus second, and now user sees that he made mistake and press back, nothing happens because we can't handle this case.  

In example bellow, just focus text field and press back on keyboard and observe that nothing is printed in console if text field is empty.

## Steps to Reproduce

```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(
    MyApp(),
  );
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  TextEditingController controller = TextEditingController();

  @override
  void initState() {
    super.initState();

    controller.addListener(() {
      print('controller text: ${controller.text}');
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: CupertinoTextField(
          controller: controller,
          placeholder: 'Click me',
          textAlign: TextAlign.center,
          onChanged: (text) {
            print('text: $text');
          },
          maxLength: 5,
        ),
      ),
    );
  }
}
```

## Logs

```
[✓] Flutter (Channel beta, v1.9.1+hotfix.4, on Mac OS X 10.14.6 18G87, locale en-US)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.0)
[✓] Android Studio (version 3.5)
[!] IntelliJ IDEA Ultimate Edition (version 2018.1.2)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] Connected device (1 available)

! Doctor found issues in 1 category.
```

**Comment by Ranjit32** [comment]
use textediting controller to listen to the changes in the text from field and then pass the data in onpressed

**Comment by fralways** [comment]
controller behaves same as onChanged. It doesn't work in this case.

**Comment by Ranjit32** [comment]
> controller behaves same as onChanged. It doesn't work in this case.

dry to store your text in another variable like changedtext = text and try  printing changedtext

**Comment by iapicca** [comment]
Hi @fralways 
if the issue persist with the latest stable version of flutter
could you please provide your updated `flutter doctor -v` ,
and your `flutter run --verbose` 
Thank you

**Comment by fralways** [comment]
@iapicca Yes it does.

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.12.13+hotfix.8, on Mac OS X 10.14.6 18G103, locale en-US)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
[✓] Android Studio (version 3.6)
[!] VS Code (version 1.43.0)
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter
[✓] Connected device (1 available)
```

When we have empty text field:
onChanged callback is not called when we press backspace. 
TextEditingController listener is not called when we press backspace. 

On other side, if we reach text limit:
onChanged callback is called
TextEditingController listener is not called

I would expect that we get callbacks for both cases.

I have updated example to include TextEditingController as well.

**Comment by pedromassangocode** [comment]
I was able to reproduce this on latest stable channel, however I'm not sure if `onChanged` should get called when the text does not actually changed. Those callbacks (`onChanged` and controller.addListener()) are called only when the data in the text field changes. 

This may be considered a proposal for a new callback or so.

<details><summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 1.20.3, on Mac OS X 10.15.6 19G2021, locale en)
    • Flutter version 1.20.3 at /Users/pedromassango/dev/SDKs/flutter_stable
    • Framework revision 216dee60c0 (9 days ago), 2020-09-01 12:24:47 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/pedromassango/Library/Android/sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.7)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.7, Build version 11E801a
    • CocoaPods version 1.9.3

[!] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Community Edition (version 2020.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 48.1.4
    • Dart plugin version 202.7206

[✓] VS Code (version 1.48.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.2

[✓] Connected device (1 available)
    • sdk gphone x86 arm (mobile) • emulator-5556 • android-x86 • Android 11 (API 30) (emulator)

! Doctor found issues in 1 category.
Process finished with exit code 0
```
</details>

**Comment by pedromassangocode** [comment]
Just for the records the native iOS callbacks can be found at:
https://developer.apple.com/documentation/uikit/uitextfielddelegate

This issue may be related to adding support for `textFieldShouldReturn` callback:
https://developer.apple.com/documentation/uikit/uitextfielddelegate/1619603-textfieldshouldreturn

**Comment by fralways** [comment]
@pedromassango following up on my previous comment onChanged callback is called even when nothing is changed (when we reach text field character limit, and user enters new character). so we have different behaviour when text field is empty and we try to delete, and when we reach limit and try to enter character. behaviours should be the same, either we get callbacks or we dont get callbacks. 

now, if flutter team decides not to have callbacks, it should be first implemented like that and only after that it makes sense to treat this as a proposal, but in my pov, we should get callbacks for both cases by default and not introduce new property.

**Comment by pedromassangocode** [comment]
> @pedromassango following up on my previous comment onChanged callback is called even when nothing is changed (when we reach text field character limit, and user enters new character)

This can be considered a bug, because the onChanged method (as the name suggests) is supposed to be called only when the text field changes. We have this as a new issue.

Cc: @justinmc who have been working on text fields widgets.

**Comment by justinmc** [comment]
I believe this is the same underlying issue as https://github.com/flutter/flutter/issues/14809, and may therefore be very hard to solve.  CC @gspencergoog

**Comment by gspencergoog** [comment]
Yes, this is a duplicate of #14809
